examples/explorer/data_feed/tycho.rs

```diff
@@ -137,29 +234,63 @@ pub async fn process_messages(
                 }
 
                 if !skip_pool {
-                    new_pairs.insert(id, ProtocolComponent::new(id, pair_tokens));
+                    new_pairs.insert(id.clone(), ProtocolComponent::new(id.clone(), pair_tokens));
                 }
 
                 let state: Box<dyn ProtocolSim> = match protocol.as_str() {
-                    "uniswap_v3" => match UniswapV3State::try_from(snapshot) {
+                    "uniswap_v3" => match UniswapV3State::try_from(snapshot.clone()) {
                         Ok(state) => Box::new(state),
                         Err(e) => {
                             debug!("Failed parsing uniswap-v3 snapshot! {} for pool {:x?}", e, id);
                             continue;
                         }
                     },
-                    "uniswap_v2" => match UniswapV2State::try_from(snapshot) {
+                    "uniswap_v2" => match UniswapV2State::try_from(snapshot.clone()) {
                         Ok(state) => Box::new(state),
                         Err(e) => {
                             warn!("Failed parsing uniswap-v2 snapshot! {} for pool {:x?}", e, id);
                             continue;
                         }
                     },
+                    "vm:balancer" => {
+                        match VMPoolState::try_from_with_block(
+                            snapshot.clone(),
+                            header.clone(),
+                            all_tokens.clone(),
+                        )
+                        .await
+                        {
+                            Ok(state) => {
+                                // Skip balancer pool if it doesn't pass the filter
+                                if !balancer_pool_filter(&snapshot) {
+                                    continue;
+                                }
+                                Box::new(state)
+                            }
+                            Err(e) => {
+                                warn!(
+                                    "Failed parsing balancer-v2 snapshot! {} for pool {:x?}",
+                                    e, id
+                                );
+                                continue;
+                            }
+                        }
+                    }
                     _ => panic!("VM snapshot not supported!"),
                 };
                 new_components.insert(id, state);
             }
-
+            let storage_by_address: HashMap<Address, ResponseAccount> = protocol_msg
+                .clone()
+                .snapshots
+                .get_vm_storage()
+                .into_iter()
+                .map(|(key, value)| (Address::from_slice(&key[..20]), value.clone().into()))
+                .collect();
+            info!("Updating engine with snapshot");
+            update_engine(SHARED_TYCHO_DB.clone(), block, Some(storage_by_address), HashMap::new())
```

Comment:
This needs to be done before creating the components. The vm storage needs to be in the engine before we actually make the pools. we are setting spot prices on creation now so we need to simulate right away. For balancer case, not updating the engine first is making the balancer vault unavailable and the simulations fail miserably